### PR TITLE
Feature/support extended proprietary policy

### PR DESCRIPTION
**Pull request by LuxoftAKutsan**
Add extended proprietary policy flow
Changes moved from F-S commit Feature/support extended proprietary policy
Partially reviewed in  #1053

### CODE REVIEW

**Comment by LuxoftAKutsan** [comment]
@Abyzhynar, @agritsevich, @okozlovlux, @dev-gh, @dtrunov, @orhan-mehmedov @AGaliuzov
please review.
Note that main commits to review are 48446e2 and 5310d1b.
Others are copy/paste or style fix

File: src/components/application_manager/src/commands/mobile/send_location_request.cc [hunk 1]
```diff
@@ -62,8 +62,7 @@ void SendLocationRequest::Run() {
 
   smart_objects::SmartObject& msg_params = (*message_)[strings::msg_params];
   if (msg_params.keyExists(strings::delivery_mode)) {
-    const std::vector<std::string>& allowed_params =
-        parameters_permissions().allowed_params;
+    const RPCParams& allowed_params = parameters_permissions().allowed_params;
     if (allowed_params.end() == std::find(allowed_params.begin(),
                                           allowed_params.end(),
```
Comment (orhan-mehmedov) [inline src/components/application_manager/src/commands/mobile/send_location_request.cc]: @LuxoftAKutsan There is in_range<>() in helpers.h. I think instead find(), == end() etc. can be more clear and shorter.

File: src/components/application_manager/src/policies/policy_handler.cc [hunk 2]
```diff
@@ -897,6 +914,45 @@ bool PolicyHandler::UnloadPolicyLibrary() {
...
+  SDLAlowedNotification(const connection_handler::DeviceHandle& device_id,
+                        PolicyManager* policy_manager,
+                        StateController& state_controller)
+      : device_id_(device_id)
+      , policy_manager_(policy_manager)
+      , state_controller_(state_controller) {}
+
+  void operator()(const ApplicationSharedPtr& app) {
+    if (!policy_manager_) {
+      return;
```
Comment (orhan-mehmedov) [inline src/components/application_manager/src/policies/policy_handler.cc]: @LuxoftAKutsan Why not DCHECK_OR_RETURN_VOID()?

File: src/components/application_manager/src/policies/policy_handler.cc [hunk 3]
```diff
@@ -897,6 +914,45 @@ bool PolicyHandler::UnloadPolicyLibrary() {
...
+    }
+    if (device_id_ == app->device()) {
+      std::string hmi_level;
+      mobile_apis::HMILevel::eType default_mobile_hmi;
+      policy_manager_->GetDefaultHmi(app->policy_app_id(), &hmi_level);
+      if ("BACKGROUND" == hmi_level) {
```
Comment (orhan-mehmedov) [inline src/components/application_manager/src/policies/policy_handler.cc]: @LuxoftAKutsan I found "BACKGROUND" almost 10 times in our project. Isn't it time for a constant? Check others below too.

File: src/components/application_manager/src/policies/policy_handler.cc [hunk 4]
```diff
@@ -977,8 +1075,44 @@ void PolicyHandler::OnActivateApp(uint32_t connection_key,
...
+
+    usage.RecordAppUserSelection();
+    DeviceParams device_params = GetDeviceParams(
+        app->device(),
+        application_manager_.connection_handler().get_session_observer());
+    permissions.deviceInfo = device_params;
+
+    DeviceConsent consent = policy_manager_->GetUserConsentForDevice(
+        permissions.deviceInfo.device_mac_address);
+    permissions.isSDLAllowed = kDeviceAllowed == consent ? true : false;
```
Comment (orhan-mehmedov) [inline src/components/application_manager/src/policies/policy_handler.cc]: @LuxoftAKutsan permissions.isSDLAllowed = kDeviceAllowed == consent; is enough.

File: src/components/application_manager/src/policies/policy_handler.cc [hunk 5]
```diff
@@ -977,8 +1075,44 @@ void PolicyHandler::OnActivateApp(uint32_t connection_key,
...
+
+    // According to the SDLAQ-CRS-2794, p.9
+    // 'priority' should be ommited in case when device
+    // is not allowed.
+    if (permissions.isSDLAllowed == false) {
```
Comment (orhan-mehmedov) [inline src/components/application_manager/src/policies/policy_handler.cc]: @LuxoftAKutsan just !permissions.isSDLAllowed. At least is used few lines below. In one or other way good to be consistent.

File: src/components/application_manager/src/policies/policy_handler.cc [hunk 6]
```diff
@@ -1292,7 +1448,84 @@ void PolicyHandler::OnUpdateHMIAppType(
...
+    return;
+  }
+  LOG4CXX_DEBUG(logger_, "Saved encrypted certificate data" << file_name);
+
+  MessageHelper::SendDecryptCertificateToHMI(file_name, application_manager_);
+}
+
+void PolicyHandler::OnEmptyCertificateArrived() const {
+  LOG4CXX_DEBUG(logger_, "Empty certificate arrived");
+  const std::string empty_certificate = "";
```
Comment (orhan-mehmedov) [inline src/components/application_manager/src/policies/policy_handler.cc]: @LuxoftAKutsan anyway default constructor will create an empty string.

File: CMakeLists.txt [hunk 7]
```diff
@@ -277,8 +281,22 @@ if (TELEMETRY_MONITOR)
     add_definitions(-DTELEMETRY_MONITOR)
 endif()
 
-if (ENABLE_EXTENDED_POLICY)
-    add_definitions(-DEXTENDED_POLICY)
+if (${EXTENDED_POLICY} STREQUAL "EXTENDED_PROPRIETARY")
+    add_definitions(-DEXTENDED_PROPRIETARY)
+    set(POLICY_PATH ${COMPONENTS_DIR}/policy/policy_premium/)
+    set(POLICY_GLOBAL_INCLUDE_PATH ${COMPONENTS_DIR}/include/policy/policy_premium/)
```
Comment (dev-gh) [inline CMakeLists.txt]: @LuxoftAKutsan 
For consistency I would be good to rename folders to 'extended_proprietary'

File: src/components/application_manager/include/application_manager/application_impl.h [hunk 8]
```diff
@@ -156,6 +156,7 @@ class ApplicationImpl : public virtual InitialApplicationDataImpl,
   void set_device(connection_handler::DeviceHandle device);
   virtual uint32_t get_grammar_id() const;
   virtual void set_grammar_id(uint32_t value);
+  bool is_audio() const OVERRIDE;
```
Comment (dev-gh) [inline src/components/application_manager/include/application_manager/application_impl.h]: @LuxoftAKutsan 
Description, FINAL ?

File: src/components/application_manager/include/application_manager/policies/policy_handler.h [hunk 9]
```diff
@@ -466,30 +483,32 @@ class PolicyHandler : public PolicyHandlerInterface,
       policy_handler_->AsyncRun(new StatisticsDelegate(*policy_handler_, type));
     }
 
-    void Increment(const std::string& app_id,
-                   usage_statistics::AppCounterId type) {
+    virtual void Increment(const std::string& app_id,
```
Comment (dev-gh) [inline src/components/application_manager/include/application_manager/policies/policy_handler.h]: @LuxoftAKutsan 
OVERRIDE

File: src/components/application_manager/include/application_manager/policies/regular/pt_exchange_handler.h [hunk 10]
```diff
@@ -0,0 +1,47 @@
+/*
```
Comment (dev-gh) [inline src/components/application_manager/include/application_manager/policies/regular/pt_exchange_handler.h]: @LuxoftAKutsan 
This file is not being used anywhere

File: src/components/application_manager/include/application_manager/policies/regular/pt_exchange_handler_ext.h [hunk 11]
```diff
@@ -0,0 +1,52 @@
+/*
```
Comment (dev-gh) [inline src/components/application_manager/include/application_manager/policies/regular/pt_exchange_handler_ext.h]: @LuxoftAKutsan 
This file is not being used anywhere

File: src/components/application_manager/include/application_manager/policies/regular/pt_exchange_handler_impl.h [hunk 12]
```diff
@@ -0,0 +1,61 @@
+/*
```
Comment (dev-gh) [inline src/components/application_manager/include/application_manager/policies/regular/pt_exchange_handler_impl.h]: @LuxoftAKutsan 
This file is not being used anywhere

File: src/components/include/application_manager/policies/policy_handler_interface.h [hunk 13]
```diff
@@ -85,8 +92,10 @@ class PolicyHandlerInterface {
   virtual bool GetInitialAppData(const std::string& application_id,
                                  StringArray* nicknames = NULL,
                                  StringArray* app_hmi_types = NULL) = 0;
-  virtual void GetServiceUrls(const std::string& service_type,
-                              EndpointUrls& end_points) = 0;
+  virtual void GetUpdateUrls(const std::string& service_type,
```
Comment (dev-gh) [inline src/components/include/application_manager/policies/policy_handler_interface.h]: @LuxoftAKutsan 
Need to separate with ifdef

File: src/components/policy/policy_premium/doc/doxygen/components/SmartObjects/Smart Objects Types/Type casts.txt [hunk 14]
```diff
@@ -0,0 +1,204 @@
+/*! \page components_smartobjects_types_cast Type cast for SmartObjects
+
```
Comment (dev-gh) [inline src/components/policy/policy_premium/doc/doxygen/components/SmartObjects/Smart Objects Types/Type casts.txt]: @LuxoftAKutsan 
How are these doxygens related to policy merge?

File: src/components/policy/policy_premium/include/policy/policy_table_interface.xml [hunk 15]
```diff
@@ -0,0 +1,230 @@
+<?xml version="1.0" standalone="no"?>
```
Comment (dev-gh) [inline src/components/policy/policy_premium/include/policy/policy_table_interface.xml]: @LuxoftAKutsan 
This file is not being used anymore, policy table structure defined statically now.

File: src/components/policy/policy_premium/include/policy/policy_table_interface_ext.xml [hunk 16]
```diff
@@ -0,0 +1,284 @@
+<?xml version="1.0" standalone="no"?>
```
Comment (dev-gh) [inline src/components/policy/policy_premium/include/policy/policy_table_interface_ext.xml]: @LuxoftAKutsan 
This file is not being used anymore.

**Comment by LuxoftAKutsan** [comment]
Review will be finished within #1069

File: src/components/application_manager/include/application_manager/policies/regular/pt_exchange_handler.h (hunk 10 above)
Comment (LuxoftAKutsan) [inline src/components/application_manager/include/application_manager/policies/regular/pt_exchange_handler.h]: @dev-gh it does 
```
grep pt_exchange_handler.h  -R ./                                                                                                                                                                                                     2 ↵
./src/components/application_manager/include/application_manager/policies/regular/pt_exchange_handler_ext.h:#include "application_manager/policies/pt_exchange_handler.h"
./src/components/application_manager/include/application_manager/policies/regular/pt_exchange_handler_impl.h:#include "application_manager/policies/pt_exchange_handler.h"

```

Comment (dev-gh) [inline src/components/application_manager/include/application_manager/policies/regular/pt_exchange_handler.h]: @LuxoftAKutsan 
Look to my next comment
